### PR TITLE
Implement `tensordot` and `repeat`

### PR DESCRIPTION
**Pull request by neNasko1**
This PR adds implementations to [tensordot](https://data-apis.org/array-api/2022.12/API_specification/generated/array_api.tensordot.html) and [repeat](https://data-apis.org/array-api/2023.12/API_specification/generated/array_api.repeat.html). 

Note they currently segfault on the latest onnx release, so the arrayapi tests are disabled. Coverage is provided instead on our side with some tests.

### CODE REVIEW

File: ndonnx/_core/_shapeimpl.py [hunk 1]
```diff
@@ -66,6 +66,25 @@ def expand_dims(self, x, axis):
             )
         )
 
+    def repeat(self, x, repeats, axis):
```
[the remark this answers is not in the document]
Comment (adityagoel4512) [reply ndonnx/_core/_shapeimpl.py:69]: For example:

> If axis is None, let N = prod(x.shape) and if repeats is an array, repeats must be broadcast compatible with the shape (N,) (i.e., be a one-dimensional array having shape (1,) or (N,)).

The rank can always be recovered, so checking if `repeats` is 1D is easy to do. (the rank strictly speaking can be lost if it is data-dependent in contrived cases, but you can recover it with reshapes if you know what you're doing)

Also very easy to check:

> If repeats is an array, the array must have an integer data type.

Some of this may possibly fail down the stack, but we aim to localize the error handling where possible so the error messages are precise.

File: ndonnx/_core/_shapeimpl.py [hunk 2]
```diff
@@ -66,6 +66,25 @@ def expand_dims(self, x, axis):
...
+            # TODO: this case can be optimized by broadcasting and reshaping
+            repeats = ndx.asarray(repeats)
+
+        repeats = ndx.broadcast_to(repeats, ndx.reshape(x_shape[axis], [1]))
+        indices = ndx.searchsorted(
+            ndx.cumulative_sum(repeats).astype(ndx.uint64),
+            ndx.arange(ndx.sum(repeats)),
+            side="right",
+        )
+        return ndx.take(x, indices, axis=axis)
```
Comment (adityagoel4512) [inline ndonnx/_core/_shapeimpl.py:86]: Nice trick! Not sure if it's what you already allude to in your TODO but I believe the cumulative sum search is only required for the case where you have different repeat values. If `repeat` is a python scalar, we can skip the very expensive `searchsorted`. We can revisit this in a future PR but it'll be useful to know how much performance we're giving up.

File: xfails.txt [hunk 3]
```diff
@@ -1,3 +1,7 @@
+# segmentation fault: https://github.com/onnx/onnx/pull/6570
+array_api_tests/test_manipulation_functions.py::test_repeat
+array_api_tests/test_linalg.py::test_tensordot
```
Comment (adityagoel4512) [inline xfails.txt:3]: I don't see what https://github.com/onnx/onnx/pull/6570 has to do with `tensordot`?

**Review by adityagoel4512 — CHANGES_REQUESTED** [review]
I'll need to play around with the einsum a little more but here's a first pass.

File: xfails.txt (hunk 3 above)
Comment (neNasko1) [reply xfails.txt:3]: After taking a good look, it seems I was mistaken - the actual segmentation fault came from the `onnxruntime` failing on `einsum` with zero dimensional input. 

Will submit a PR later, but in the meantime I improved the testing of `tensordot` to include some parametrised tests.

File: tests/test_core.py [hunk 4]
```diff
@@ -9,6 +9,9 @@
 import numpy as np
 import pytest
 import spox.opset.ai.onnx.v19 as op
+from hypothesis import given
```
Comment (cbourjau) [inline tests/test_core.py:12]: Let's keep hypothesis out of the ndonnx test suite for the time being, please. Would you mind summarizing what this test has that is lacking upstream? Does it make sense to upstream it?

File: xfails.txt [hunk 5]
```diff
@@ -1,3 +1,8 @@
+# segmentation fault: https://github.com/onnx/onnx/pull/6570
+array_api_tests/test_manipulation_functions.py::test_repeat
+# segmentation fault: due to onnxruntime not handling einsum with dim 0
+array_api_tests/test_linalg.py::test_tensordot
```
Comment (cbourjau) [inline xfails.txt:4]: If these tests segfault then we should put them into skips.txt rather than here. That the array-api-tests are green nonetheless is an indication that something else is fishy. That feeling is confounded by the fact that they finish within 20s. Could you take a look, please?

File: xfails.txt (hunk 3 above)
Comment (neNasko1) [reply xfails.txt:4]: > failing on einsum with zero dimensional input
The problem is in a shape, which has a zero dimension, not a shape that is with zero dimensions. E.g. `(0, 0, 1, 1)` is segfaulting, `()` is not.

Sorry for not being clear!

File: tests/test_core.py (hunk 4 above)
Comment (neNasko1) [reply tests/test_core.py:12]: The upstream is not missing anything, since we need to skip the test however @adityagoel4512 suggested we test ourselves, as to be sure that tensordot works correctly.

Comment (adityagoel4512) [reply tests/test_core.py:12]: I was commenting on the fact that a few edge cases weren't included in our tests. I don't think we need to go to property-based tests for this.

Comment (neNasko1) [reply tests/test_core.py]: Removed the parametrised testing and included some more diligent testing.

File: tests/test_core.py [hunk 6]
```diff
@@ -1038,3 +1038,83 @@ def test_argmaxmin_unsupported_kernels(func, x):
...
+        (np.arange(60).reshape(3, 4, 5), 3, 1),
+        (np.arange(60).reshape(3, 4, 5), 3, 2),
+        (np.arange(60).reshape(3, 4, 5), 3, None),
+        (np.arange(60).reshape(3, 4, 5), np.arange(3), 0),
+        (np.arange(60).reshape(3, 4, 5), np.arange(4), 1),
+        (np.arange(60).reshape(3, 4, 5), np.arange(5), 2),
+        (np.arange(60).reshape(3, 4, 5), np.arange(60), None),
+    ],
+)
+def test_repeat(a, repeats, axis, lazy_repeats):
```
Comment (adityagoel4512) [inline tests/test_core.py:1102]: What is "lazy" about the `lazy_repeats` case here? Even when `repeats` is `int`, it's immediately wrapped in `ndx.asarray(repeats)` in your implementation of `repeat`.

Comment (adityagoel4512) [reply tests/test_core.py:1102]: For a test that exercises shape inference for `repeat`, you could construct an ONNX model using `ndx.build` with a lazy `repeat` input.

Comment (neNasko1) [reply tests/test_core.py]: The idea of the test is to test that if we actually supply the argument as a python integer, we wrap it correctly. I rewrote the test to be cleaner.

Comment (adityagoel4512) [reply tests/test_core.py]: @neNasko1 please rebase

File: skips.txt [hunk 7]
```diff
@@ -2,3 +2,9 @@
 
 array_api_tests/test_manipulation_functions.py::test_roll
 array_api_tests/test_data_type_functions.py::test_broadcast_arrays
+
+# segmentation fault: https://github.com/onnx/onnx/pull/6570
+array_api_tests/test_manipulation_functions.py::test_repeat
+
+# segmentation fault: due to onnxruntime not handling einsum with dim 0
```
Comment (adityagoel4512) [inline skips.txt:9]:
```suggestion
# segmentation fault: https://github.com/microsoft/onnxruntime/pull/23379
```

File: tests/test_core.py [hunk 8]
```diff
@@ -1045,6 +1045,88 @@ def test_argmaxmin_unsupported_kernels(func, x):
...
+    [
+        (np.arange(60).reshape(3, 4, 5), np.arange(60).reshape([3, 4, 5]), 0),
+        (np.arange(60).reshape(3, 4, 5), np.arange(60).reshape([3, 20]), None),
+        (np.arange(60).reshape(3, 4, 5), np.arange(60).reshape([2, 3, 2, 5]), None),
+    ],
+)
+def test_repeat_raises(a, repeats, axis):
+    with pytest.raises(ValueError):
+        ndx.repeat(ndx.asarray(a), repeats, axis=axis).to_numpy()
+
```
Comment (adityagoel4512) [inline tests/test_core.py:1129]:
```suggestion


```

**Comment by neNasko1** [comment]
Thanks for merging into main!

**Review by adityagoel4512 — APPROVED** [review]
